// spicetify-creator/dist/buildCustomApp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const glob_1 = __importDefault(require("glob"));
const chalk_1 = __importDefault(require("chalk"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const extractFiles_1 = __importDefault(require("./helpers/extractFiles"));
const minify_1 = require("./helpers/minify");
const esbuild_1 = __importDefault(require("esbuild"));
const os_1 = __importDefault(require("os"));
exports.default = async (settings, outDirectory, watch, esbuildOptions, minify, inDirectory) => {
    const extensions = glob_1.default.sync(`${inDirectory}/extensions/*(*.ts|*.tsx|*.js|*.jsx)`);
    const extensionsNewNames = extensions.map(e => e.substring(0, e.lastIndexOf(".")) + ".js");
    const iconPath = settings.icon ? path_1.default.join(inDirectory, settings.icon) : null;
    const activeIconPath = settings.activeIcon ? path_1.default.join(inDirectory, settings.activeIcon) : iconPath;
    console.log("Generating manifest.json...");
    const customAppManifest = {
        name: settings.displayName,
        icon: iconPath ? fs_1.default.readFileSync(iconPath, 'utf-8') : "",
        "active-icon": activeIconPath ? fs_1.default.readFileSync(activeIconPath, 'utf-8') : "",
        subfiles: [],
        subfiles_extension: extensionsNewNames.map(e => path_1.default.basename(e))
    };
    fs_1.default.writeFileSync(path_1.default.join(outDirectory, "manifest.json"), JSON.stringify(customAppManifest, null, 2));
    const appPath = path_1.default.resolve(glob_1.default.sync(`${inDirectory}/*(app.ts|app.tsx|app.js|app.jsx)`)[0]);
    const tempFolder = path_1.default.join(os_1.default.tmpdir(), "spicetify-creator");
    const indexPath = path_1.default.join(tempFolder, `index.jsx`);
    if (!fs_1.default.existsSync(tempFolder))
        fs_1.default.mkdirSync(tempFolder);
    fs_1.default.writeFileSync(indexPath, `
import App from \'${appPath.replace(/\\/g, "/")}\'
import React from 'react';

export default function render() {
  return <App />;
}
  `.trim());
    esbuild_1.default.build(Object.assign(Object.assign({ entryPoints: [indexPath, ...extensions], outdir: outDirectory }, esbuildOptions), { watch: (watch ? {
            async onRebuild(error, result) {
                if (error)
                    console.error(error);
                else {
                    await afterBundle();
                }
            },
        } : undefined) })).then(async (r) => {
        await afterBundle();
        return r;
    });
    const afterBundle = async () => {
        console.log("Moving files out of folders...");
        (0, extractFiles_1.default)(outDirectory, true);
        console.log("Modifying index.js...");
        fs_1.default.appendFileSync(path_1.default.join(outDirectory, "index.js"), `const render=()=>${esbuildOptions.globalName}.default();\n`);
        console.log("Renaming index.css...");
        if (fs_1.default.existsSync(path_1.default.join(outDirectory, "index.css")))
            fs_1.default.renameSync(path_1.default.join(outDirectory, "index.css"), path_1.default.join(outDirectory, "style.css"));
        // Account for dynamic hooking of React and ReactDOM
        extensionsNewNames.map(e => path_1.default.basename(e)).map(extensionFile => {
            const extensionFilePath = path_1.default.join(outDirectory, extensionFile);
            fs_1.default.writeFileSync(extensionFilePath, `
        (async function() {
          while (!Spicetify.React || !Spicetify.ReactDOM) {
            await new Promise(resolve => setTimeout(resolve, 10));
          }
          ${fs_1.default.readFileSync(extensionFilePath, "utf-8")}
        })();
      `.trim());
        });
        if (minify) {
            console.log("Minifying...");
            await (0, minify_1.minifyFolder)(outDirectory);
        }
        console.log(chalk_1.default.green('Build succeeded.'));
    };
};
